### ISSUE TITLE
Maintain xhprof and make it a cool and maintained tool in 2021

### ISSUE THREAD

**Issue by williamdes** [code-lts/welcome#6]
Hi everybody,

The xhprof project and tool are very very cool !

Today 2 repositories share the xhprof name and I would like to try to unite all our forces to have the tool be clean and maintained. ([find active forks](https://techgaun.github.io/active-forks/index.html#phacility/xhprof))

The first repository is https://github.com/longxinH/xhprof and @longxinH and contributors have done an awesome work having the PECL extension PHP 7 and 8 compatible :rocket: :100: 

And the second repository is  @preinheimer 's https://github.com/preinheimer/xhprof that has also quite a lot of stars

Open issue for maintenance at `longxinH/xhprof`: https://github.com/longxinH/xhprof/issues/52

This issue refers to `project status` in https://github.com/phacility/xhprof/issues/85 (not updated since 2016)

Also this repository (https://github.com/humanmade/xhprof/pulls) has 2 PRs made by @joehoyle and I feed they should be open into @longxinH 's fork.

NB: [tideways/php-xhprof-extension](https://github.com/tideways/php-xhprof-extension) seems to be a maintained project afternative

-----

My proposal would be to get back the original projet by https://github.com/phacility/xhprof if @epriestley agrees to transfer it here.

Or to maintain forks of the project but improve what we are all doing by creating a website or something that would make the project seem to be clean and up to date.

Having a docker image for the UI would be great too.

The UI needs some maintenance because it has got none since 2014 (if I refer to `longxinH/xhprof`).

People that would like to be notified, cc list: @Guileas @epriestley @longxinH @carlwgeorge @aik099 @joehoyle @beberlei

----

Please comment and let me know what you think about all that :)

**Comment by joehoyle** [comment]
> Also this repository (https://github.com/humanmade/xhprof/pulls) has 2 PRs made by @joehoyle and I feed they should be open into @longxinH 's fork.

That's right -- we use XHProf only in sampling mode, (hence https://github.com/humanmade/xhprof/pull/2), as that's the only one really suitable for production profiling. Sampling wasn't brought over to the Tideways fork, which is why that's not an alternative for us at least.

I also recently added ARM support for Apple M1, though I really barely know what I'm doing when it comes to C++/C so this was based off some Googling and trial by error.

It's interesting timing that you brought this up -- I had all but given up on XHProf, and we've ben considering moving to https://www.mediawiki.org/wiki/Excimer. Wikimedia built this extension off the back of XHProf being a bit of a dead project, and as far as I know they only used it for the same purpose we did: gathering performance profile with time samples data and build flame graphs in an external UI.

Given that Excimer is much simpler and more specific to our needs, I think it will make sense for us to still migrate to that project, and therefor I don't anticipate much contribution and sponsorship from us -- just to be totally transparent and set expectations!

**Comment by preinheimer** [comment]
I am supportive of any efforts to make the tool easy to use and am willing to transfer somewhere if required.

I also worked on a similar tool: https://github.com/perftools/xhgui which was transferred to an organization https://github.com/perftools to make ownership and administration easier.

**Comment by williamdes** [comment]
> I am supportive of any efforts to make the tool easy to use and am willing to transfer somewhere if required.
> 
> I also worked on a similar tool: https://github.com/perftools/xhgui which was transferred to an organization https://github.com/perftools to make ownership and administration easier.

Thanks a lot !

> and am willing to transfer somewhere if required.

Could you transfer it (https://github.com/preinheimer/xhprof) here and we could rename it afterwards to allow other forks to be migrated here ?
Afterwards I will give you rights on the repository.

**Comment by williamdes** [comment]
Update: there will be no transfer of the original repository to @code-lts
Quote (https://github.com/phacility/xhprof/issues/85#issuecomment-766916573): "No. I'd encourage you to fork it (or work with one of the existing forks) if you aren't satisfied with how it is currently being maintained."

And the official upstream project status from 2016: https://github.com/phacility/xhprof/issues/85#issuecomment-196489773

That said @longxinH would you agree to transfer your repository to @code-lts ?

**Comment by preinheimer** [comment]
@williamdes are you prepared to commit to some level of work and maintenance if someone is willing to transfer a repo?

I'm asking as there have been some high-profile mis-uses of repos over the past few years.

**Comment by williamdes** [comment]
> @williamdes are you prepared to commit to some level of work and maintenance if someone is willing to transfer a repo?
> 
> I'm asking as there have been some high-profile mis-uses of repos over the past few years.

Sure, I am willing to do some maintenance
At least on all the languages I am competent on and rely onto contributions on the ones I am not an expert on.
You can have a look to the [Doctum](https://github.com/code-lts/doctum#readme) project is you know the tool Sami. I do maintenance and improvements to maintain the tool.

**Comment by longxinH** [comment]
> That said @longxinH would you agree to transfer your repository to @code-lts ?

No, I hope to follow the original author, and the updated UI can be submitted to me in the form of PR

**Comment by aik099** [comment]
@williamdes , as a https://github.com/preinheimer/xhprof maintainer, I like the idea. I think there should be a decent free alternative for a profiling tool, that:

* you don't have to pay every month to use like https://www.blackfire.io/, https://tideways.com/ 
* has a decent amount of features
* has support for all PHP versions
* have modern UI

Collecting profiling info is good, but I mostly use it with a combination of nice UI of https://github.com/preinheimer/xhprof, which recently got an update. Also having PHP 5.x support completely dropped from the XHProf PHP extension isn't a good idea, because I have some projects on PHP 5, which are too expensive to migrate to PHP 7.

**Comment by williamdes** [comment]
Thank you for the feedback !

> I think there should be a decent free alternative for a profiling tool, that: you don't have to pay every month to use like https://www.blackfire.io/, https://tideways.com/

I support this idea ! :100: 

The pecl extension dropped PHP 5 support I think but I am not sure. Anyway I also agree but would argue that the UI can require PHP 7 but can also support some PHP 5 versions

I will be very happy to welcome your fork here at @code-lts and create a subdomain xhprof.long-term.support.

This will be a good façade for making this tool find back some audience.

@longxinH what would you think about keeping your fork for the dedicated pecl extension code and dropping the UI code because another fork already has done the work and maintains it ?

**Comment by longxinH** [comment]
> The pecl extension dropped PHP 5 support I think but I am not sure. Anyway I also agree but would argue that the UI can require PHP 7 but can also support some PHP 5 versions

I don’t think the php5 version has the meaning of continuing to support it, because now the mainstream version is in php7

https://github.com/preinheimer/xhprof It's a great UI. Do you have any suggestions to merge the two versions together? @williamdes

**Comment by williamdes** [comment]
@mailoman owns https://github.com/xhprof
Maybe we could use the @code-lts organization and have the 2 repositories live in it. Or create another organization for the xhprof tool. 
One renamed to xhprof-ui and the other one xhprof-extension

Last solution:
Merge one repository in the other one and mark the second one as transfered
But that would make one repository abandoned and all starts "lost"

Let me know if you find a better idea that suits you

**Comment by mailoman** [comment]
> @mailoman owns https://github.com/xhprof
> Maybe we could use the @code-lts organization and have the 2 repositories live in it. Or create another organization for the xhprof tool.
> One renamed to xhprof-ui and the other one xhprof-extension
> 
> Last solution:
> Merge one repository in the other one and mark the second one as transfered
> But that would make one repository abandoned and all starts "lost"
> 
> Let me know if you find a better idea that suits you

Guys, if you have any questions - just ask me directly

**Comment by williamdes** [comment]
@mailoman would you give us the xhprof organization?
Or any similar solution?

**Comment by longxinH** [comment]
@code-lts creates a new project library xhprof-ui, I submit extension code, @mailoman submits UI related code, how?

**Comment by williamdes** [comment]
> @code-lts creates a new project library xhprof-ui, I submit extension code, @mailoman submits UI related code, how?

I do nit create anything, In order to keep starts please go to your repository settings and click transfer and write "code-lts"

But if @mailoman gives me/us the organization xhprof we could use it to transfer our repositories to it

**Comment by mailoman** [comment]
I can give access to the organization if you would like

**Comment by williamdes** [comment]
> I can give access to the organization if you would like

Can you give me access, admin or owner ?

**Comment by mailoman** [comment]
ok, just sent invitation to **williamdes**

**Comment by mailoman** [comment]
let me know if any have any problem, cc @williamdes

**Comment by williamdes** [comment]
> let me know if any have any problem, cc @williamdes 

Thank you, it worked

@preinheimer would you transfer your repository to the @xhprof organization?

And @longxinH would you also want to do the same?

Between this two hypothetical transfers the repositories would be renamed to "ui" and "extension" 💡

All the stars, issues, forks, owners would be kept

**Comment by longxinH** [comment]
@williamdes @mailoman If I want to publish the code to pecl, it will be a problem to bring the new UI up, because there are some strong dependencies that will increase the complexity of use.

**Comment by williamdes** [comment]
> @williamdes @mailoman If I want to publish the code to PECL, it will be a problem to bring the new UI up, because there are some strong dependencies that will increase the complexity of use.

I am not sure to understand but I never meant to publish the UI code to pecl

On pecl will stay the actual work you did

There is better ways to install UIs in 2021 ;)

**Comment by aik099** [comment]
@williamdes pay attention, when creating `xhprof/extension` repo from @longxinH fork you'll end up in PHP extension, that only works in PHP 7 and not PHP 5 as `preinheimer/xhprof`.

**Comment by williamdes** [comment]
> @williamdes pay attention, when creating `xhprof/extension` repo from @longxinH fork you'll end up in PHP extension, that only works in PHP 7 and not PHP 5 as `preinheimer/xhprof`.

If we are talking about having an UI only compatible with PHP 7, then okay.

But if you are saying that the `preinheimer/xhprof` repository also has an extension support for PHP 5 then I would ask to @longxinH if it is possible to have PHP 5 support on the PECL extension.

**Comment by aik099** [comment]
> But if you are saying that the preinheimer/xhprof repository also has an extension support for PHP 5 then I would ask to @longxinH if it is possible to have PHP 5 support on the PECL extension.

@williamdes , it has such PHP extension, but no automated tests to prove, that it still works.

**Comment by williamdes** [comment]
Could the users having repositories start transferring them to @xhprof org ?
It would be great to start deciding about where the code is and after that discuss about plans to re-organize the code and have a website

**Comment by preinheimer** [comment]
If things are going to be moved, https://github.com/perftools/ already contains several tools that use the XHProf type data.

**Comment by aik099** [comment]
I've just checked https://pecl.php.net/package/xhprof and somehow the extension from there can be installed on every existing PHP version and is coming from https://github.com/longxinH/xhprof. Definitely it does work on PHP 5.x already, but https://github.com/longxinH/xhprof repo doesn't tell that in README.md.

**Comment by glensc** [comment]
[In reply to @aik099](https://github.com/code-lts/welcome/issues/6#issuecomment-769400777):
> I think there should be a decent free alternative for a profiling tool, that:
> 
> * you don't have to pay every month to use like [blackfire.io](https://www.blackfire.io/), [tideways.com](https://tideways.com/)
> * has a decent amount of features
> * has support for all PHP versions
> * have modern UI
> 
> Collecting profiling info is good, but I mostly use it with a combination of nice UI of [preinheimer/xhprof](https://github.com/preinheimer/xhprof) ([![](chrome-extension://bcnkgcoohaaaclieohdlkphgfinkgbfm/link-icon.png) hace 4 días](https://github.com/preinheimer/xhprof/tree/a46e7f45b14468ddd60c50dc4362c4df47464ee3)), which recently got an update. Also having PHP 5.x support completely dropped from the XHProf PHP extension isn't a good idea, because I have some projects on PHP 5, which are too expensive to migrate to PHP 7.


_As a maintainer to https://github.com/perftools/xhgui and author to https://github.com/perftools/php-profiler I can throw in some cheap promote:_

Projects in @perftools offer this, support of profiling on all PHP versions and the UI is quite modern:

- The GUI, recommended deployment is Docker: https://github.com/perftools/xhgui
- The profiler, frontend to xhprof/tideways/uprofiler/..., built for PHP 5.3+: https://github.com/perftools/php-profiler

This architecture can profile any PHP versions in production, but still have modern PHP to render and analyze the data.

I've used that setup on 5.3 recently in production.

**Comment by glensc** [comment]
> I've just checked [pecl.php.net/package/xhprof](https://pecl.php.net/package/xhprof) and somehow the extension from there can be installed on every existing PHP version and is coming from [longxinH/xhprof](https://github.com/longxinH/xhprof). Definitely it does work on PHP 5.x already, but [longxinH/xhprof](https://github.com/longxinH/xhprof) repo doesn't tell that in README.md.

pecl.php.net has copies of repositories, uploaded as tar.gz, so you can just build using archived versions:
- https://github.com/perftools/php-profiler#xhprof

**Comment by williamdes** [comment]
Hi everybody :wave: ,
Did someone decide if their repository could be transferred to @xhprof ?

I would like this tool to look nice and working and maybe have a website built for it. But IMO it is not individual repositories acting as forks that will meet this goal (maybe I am mistaken). I say that because it just feels everyone is walking in a different path.
If someone if willing to make things move and help the tool run better please say something :)

:arrow_right:  The goals:
- Have a website to link and explain how to install the extension and show screenshots of an UI
- Have an UI that works out of the box, no MySQL or other to install
- Maybe also link better UIs with more functionalities
- Be sure the pecl extension covers as much PHP versions as possible
- Know who has build alternatives, list them on the website or help them
- When everything is fine and running make sure the basic UI and advanced UI can be installed and used easily by our users

I feel there is a chance for xhprof to be a more structured project and be usable by users. 

:arrow_right: If you have thoughts let me know, if you want to move forward please also say it.

cc @andypost

**Comment by andypost** [comment]
I thing it would be great to resurrect PECL xhprof, meantime I see 71 extension already hosted in https://github.com/php?q=pecl-

Thoughts about "UI-facelift" better to keep out of making it much dependable from the extension.

PS: Did not check yet but there's great article about new PHP 8 Obeserver API which would be great to adopt looking at
- https://www.datadoghq.com/blog/engineering/php-8-observability-baked-right-in/
- https://github.com/tideways/php-xhprof-extension/pull/96

**Comment by glensc** [comment]
@andypost pecl/xhprof has been resurrected:
- https://pecl.php.net/package/xhprof

new maintainer at:
- https://github.com/longxinH/xhprof

I totally agree UI and extensions should live different life:
- extension: https://github.com/longxinH/xhprof
- UI: https://github.com/preinheimer/xhprof

ps: https://github.com/xhprof seems to be free at the moment.

**Comment by mail0man** [comment]
> ps: https://github.com/xhprof seems to be free at the moment.

No, it's not. just owner account was flagged - that's why it's not public now. Owner is contacting GH support now to resolve it AFAIK.

**Comment by andypost** [comment]
@glensc I mean that pecl could use php organisation as other 71 extensions (in terms of place to collaborate). The only problem with xhgui I know is extra storage (mongo/mysql) and conversion script which is primarily blocker for adoption

**Comment by williamdes** [comment]
> @glensc I mean that pecl could use php organisation as other 71 extensions (in terms of place to collaborate). The only problem with xhgui I know is extra storage (mongo/mysql) and conversion script which is primarily blocker for adoption

Good idea, that would be a better place

And for xhui I agree that it should not depend on a db, the original ui works great but requires a php 5 to run it.

**Comment by aik099** [comment]
@williamdes , how you can make XHProf GUI (general GUI, not the project name here) not depend on the database when you need to collect data at someplace and then use UI to show it? You can use https://github.com/preinheimer/xhprof with SQLite database if I'm not mistaken.

**Comment by williamdes** [comment]
> @williamdes , how you can make XHProf GUI (general GUI, not the project name here) not depend on the database when you need to collect data at someplace and then use UI to show it? You can use https://github.com/preinheimer/xhprof with SQLite database if I'm not mistaken.

Just depend on files like the original UI does :)
https://github.com/phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof

**Comment by aik099** [comment]
> Just depend on files like the original UI does :)
https://github.com/phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof

Interesting idea to store profiling information in a database-independent format and then let GUI to manage those files. It could be a nice addition to https://github.com/preinheimer/xhprof to support such a way of working.

**Comment by williamdes** [comment]
> > Just depend on files like the original UI does :)
> https://github.com/phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof
> 
> Interesting idea to store profiling information in a database-independent format and then let GUI to manage those files. It could be a nice addition to https://github.com/preinheimer/xhprof to support such a way of working.

Yes, and this is important because I can ask people to send me their files. 
They have nothing to install and I just can save the file they sent me and have a look to the profile.
This is awesome for debugging people reporting slowness.

**Comment by glensc** [comment]
> > > Just depend on files like the original UI does :)
> > > [phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof](https://github.com/phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof)
> > 
> > 
> > Interesting idea to store profiling information in a database-independent format and then let GUI to manage those files. It could be a nice addition to [preinheimer/xhprof](https://github.com/preinheimer/xhprof) to support such a way of working.
> 
> Yes, and this is important because I can ask people to send me their files.
> They have nothing to install and I just can save the file they sent me and have a look to the profile.
> This is awesome for debugging people reporting slowness.

Seems like "file saver" usage:
- https://github.com/perftools/php-profiler#file-saver

**Comment by andypost** [comment]
I'm maintaining https://www.drupal.org/project/xhprof

Btw as Drupal using composer last years would be great to extract base classes into some library https://git.drupalcode.org/project/xhprof/-/tree/8.x-1.x/src/XHProfLib

**Comment by williamdes** [comment]
> I'm maintaining https://www.drupal.org/project/xhprof
> 
> Btw as Drupal using composer last years would be great to extract base classes into some library https://git.drupalcode.org/project/xhprof/-/tree/8.x-1.x/src/XHProfLib

The code looks clean, some readme hints on how to run it would be helpfull

Seems like more I wait for feedback more I find people that built something cool out of the primary idea. And also that nobody feels like moving :thinking:. Best would maybe be a website to give other devs not knowing what xhprof is some hints on how to use it. And what version of the tool will fit them best. Like a theme gallery but for xhprof UIs and extensions.
